Intra-node GPU messaging #2: Additions to Converse and Charm++ layer RDMA

minitu opened the pull request:
Follow-up to #2896, second PR out of 4 PRs for intra-node GPU messaging.

- Added `CMK_ZC_DEVICE_MSG` header to discern GPU messages, and subsequently changed `CmiUInt1 zcMsgType:3` to `zcMsgType:4`
- Function & data structure implementations in Converse and Charm++ layer RDMA files for managing transfers

PR sequence: #2896 -> #2901 -> #2903 -> #2904

nitbhat commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+}
+*/
+
+static void findFreeIpcEvents(DeviceManager* dm, const size_t comm_offset, int& event_idx) {
```
Is there a particular reason you're not returning `event_idx` using the function, instead of having it be written using the reference in the function parameter list?

minitu replied:
I don't know why I did that, will update

minitu commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+      case CkNcpyModeDevice::MEMCPY:
+        {
+          // Directly invoke memcpy from source buffer to destination buffer
+          hapiCheck(cudaMemcpyAsync((void*)dest.ptr, source.ptr, std::min(source.cnt, dest.cnt),
```
Good point, will update

minitu commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+
+          // 2. Invoke cudaMemcpyAsync (from source device comm buffer to destination buffer)
+          hapiCheck(cudaMemcpyAsync((void*)dest.ptr, (void*)((char*)device_info.buffer + source.comm_offset),
+                std::min(source.cnt, dest.cnt), cudaMemcpyDeviceToDevice, postStructs[i].cuda_stream));
```
Same

minitu commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+    // TODO: Use GPUDirect RDMA
+    for (int i = 0; i < numops; i++) {
+      // Allocate temporary host buffer and copy the source buffer
+      buffers[i]->data_stored = true;
```
Exactly, this is just a stopgap mechanism for inter-node messages until GPUDirect RDMA is added.

nitbhat replied:
Okay, sounds good.

minitu commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+}
+
+// Performs sender-side operations necessary for device zerocopy
+void CkRdmaDeviceOnSender(int dest_pe, int numops, CkDeviceBuffer** buffers) {
```
Yeah it's called on the sender in the charmxi-generated code (#2903). I'll add comments explaining the overall control flow as you suggested.

minitu commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+      hapiCheck(cudaMallocHost(&buffers[i]->data, buffers[i]->cnt));
+      hapiCheck(cudaMemcpyAsync(buffers[i]->data, buffers[i]->ptr, buffers[i]->cnt,
+            cudaMemcpyDeviceToHost, buffers[i]->cuda_stream));
+      hapiCheck(cudaStreamSynchronize(buffers[i]->cuda_stream));
```
Yes, great point

minitu commented on src/ck-core/ckrdma.C:
```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+        }
+      case CkNcpyModeDevice::RDMA:
+        {
+          // Transfer the received/unpacked data on host to the destination device buffer
```
👍

minitu commented:
@nitbhat Addressed your comments, let me know if they're alright. I'll move the GPU related code to a different file (i.e. `ckrdmadevice.C`) and add high-level comments there afterwards.

nitbhat commented:
> @nitbhat Addressed your comments, let me know if they're alright. I'll move the GPU related code to a different file (i.e. `ckrdmadevice.C`) and add high-level comments there afterwards.

Yes, looks good now. That sounds good. Thanks!

minitu commented:
@nitbhat Moved the code and added some high-level comments in `ckrdmadevice.C`. Please let me know what you think.

nitbhat commented on src/conv-core/conv-rdmadevice.h:
```diff
@@ -0,0 +1,75 @@
+  int device_idx;
+  size_t comm_offset;
+  int event_idx;
+  cudaStream_t cuda_stream;
```
Just confirming that it is not required to pup cuda_stream?

minitu replied:
Correct, `cuda_stream` in `CmiDeviceBuffer` is only used on the sender side in `CkRdmaDeviceOnSender` in `ckrdmadevice.C`.
On the receiver side the `cuda_stream` from the user-specified `CkDeviceBufferPost` is used, in `CkRdmaIssueRgets` which is also in `ckrdmadevice.C`.
Thanks for checking on this though.

nitbhat commented:
@minitu: This patch should be good to go if you fix the compilation errors showing up in netlrts linux cuda ci build
```

2020-07-04T03:58:13.3311184Z /bin/bash ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../.. -I../../hwloc/include/private/autogen -I../../hwloc/include/hwloc/autogen  -I/home/runner/work/charm/charm/netlrts-linux-x86_64-cuda/tmp/hwloc/include -I/home/runner/work/charm/charm/netlrts-linux-x86_64-cuda/tmp/hwloc/include -DHWLOC_INSIDE_LIBHWLOC -DHWLOC_PLUGINS_PATH="\"/usr/local/lib/hwloc\"" -DRUNSTATEDIR="\"\""  -I./../include -I/usr/include -D__CHARMC__=1 -fno-stack-protector     -MT bitmap.lo -MD -MP -MF .deps/bitmap.Tpo -c -o bitmap.lo bitmap.c
2020-07-04T03:58:13.5080237Z init.C: In function ‘void _initCharm(int, char**)’:
2020-07-04T03:58:13.5084292Z init.C:1628:7: error: ‘hapiInitCsv’ was not declared in this scope
2020-07-04T03:58:13.5085485Z        hapiInitCsv(); // Initialize per-process variables (GPUManager)
2020-07-04T03:58:13.5085874Z        ^~~~~~~~~~~
2020-07-04T03:58:13.5140390Z init.C:1628:7: note: suggested alternative: ‘hapiInit’
2020-07-04T03:58:13.5142553Z        hapiInitCsv(); // Initialize per-process variables (GPUManager)
2020-07-04T03:58:13.5143057Z        ^~~~~~~~~~~
2020-07-04T03:58:13.5143401Z        hapiInit
2020-07-04T03:58:13.5145399Z init.C:1630:5: error: ‘hapiInitCpv’ was not declared in this scope
2020-07-04T03:58:13.5146146Z      hapiInitCpv(); // Initialize per-PE variables
2020-07-04T03:58:13.5146552Z      ^~~~~~~~~~~
2020-07-04T03:58:13.5181195Z init.C:1630:5: note: suggested alternative: ‘hapiInit’
2020-07-04T03:58:13.5182084Z      hapiInitCpv(); // Initialize per-PE variables
2020-07-04T03:58:13.5182463Z      ^~~~~~~~~~~
2020-07-04T03:58:13.5182849Z      hapiInit
2020-07-04T03:58:13.5189440Z init.C:1634:5: error: ‘hapiMapping’ was not declared in this scope
2020-07-04T03:58:13.5190183Z      hapiMapping(argv); // Perform PE-device mapping
2020-07-04T03:58:13.5190584Z      ^~~~~~~~~~~
2020-07-04T03:58:13.5255717Z init.C:1634:5: note: suggested alternative: ‘Mapping’
2020-07-04T03:58:13.5257275Z      hapiMapping(argv); // Perform PE-device mapping
2020-07-04T03:58:13.5258276Z      ^~~~~~~~~~~
2020-07-04T03:58:13.5258655Z      Mapping
2020-07-04T03:58:13.5261501Z init.C:1642:7: error: ‘shmCreate’ was not declared in this scope
2020-07-04T03:58:13.5262171Z        shmCreate(); // Create a per-host shared memory region
2020-07-04T03:58:13.5263053Z        ^~~~~~~~~
2020-07-04T03:58:13.5283804Z init.C:1642:7: note: suggested alternative: ‘QdCreate’
2020-07-04T03:58:13.5284482Z        shmCreate(); // Create a per-host shared memory region
2020-07-04T03:58:13.5284867Z        ^~~~~~~~~
2020-07-04T03:58:13.5285171Z        QdCreate
2020-07-04T03:58:13.5287103Z init.C:1647:5: error: ‘ipcHandleCreate’ was not declared in this scope
2020-07-04T03:58:13.5288086Z      ipcHandleCreate(); // Create CUDA IPC handles
2020-07-04T03:58:13.5288204Z      ^~~~~~~~~~~~~~~
2020-07-04T03:58:13.5354197Z init.C:1647:5: note: suggested alternative: ‘CcsHandlerTable’
2020-07-04T03:58:13.5354677Z      ipcHandleCreate(); // Create CUDA IPC handles
2020-07-04T03:58:13.5355006Z      ^~~~~~~~~~~~~~~
2020-07-04T03:58:13.5355293Z      CcsHandlerTable
2020-07-04T03:58:13.5357898Z init.C:1657:7: error: ‘ipcHandleOpen’ was not declared in this scope
2020-07-04T03:58:13.5358625Z        ipcHandleOpen(); // Open CUDA IPC handles for accessing other processes' device memory
2020-07-04T03:58:13.5359229Z        ^~~~~~~~~~~~~
2020-07-04T03:58:13.5399434Z init.C:1657:7: note: suggested alternative: ‘CcsHandlerRec’
2020-07-04T03:58:13.5400291Z        ipcHandleOpen(); // Open CUDA IPC handles for accessing other processes' device memory
2020-07-04T03:58:13.5401200Z        ^~~~~~~~~~~~~
2020-07-04T03:58:13.5401314Z        CcsHandlerRec
2020-07-04T03:58:13.5411838Z init.C:1661:5: error: ‘hapiRegisterCallbacks’ was not declared in this scope
2020-07-04T03:58:13.5412287Z      hapiRegisterCallbacks();
2020-07-04T03:58:13.5412613Z      ^~~~~~~~~~~~~~~~~~~~~
2020-07-04T03:58:13.5473261Z init.C:1661:5: note: suggested alternative: ‘CQdRegisterCallback’
2020-07-04T03:58:13.5473745Z      hapiRegisterCallbacks();
2020-07-04T03:58:13.5474182Z      ^~~~~~~~~~~~~~~~~~~~~
2020-07-04T03:58:13.5474512Z      CQdRegisterCallback
2020-07-04T03:58:13.7540023Z ../bin/charmc   -I.   -c -o qd.o qd.C
```

minitu commented:
Just to clarify, the device API differs from the host API in that the regular entry method is invoked only after the data transfers are **initiated**, not completed. This is to facilitate asynchrony by allowing the user to offload other GPU work that is dependent on the arriving data without having to explicitly synchronize.

stwhite91 commented on src/ck-core/ckrdmadevice.C:
```diff
@@ -0,0 +1,300 @@
+
+// Performs sender-side operations necessary for device zerocopy
+void CkRdmaDeviceOnSender(int dest_pe, int numops, CkDeviceBuffer** buffers) {
+  // TODO: Need to handle the case where the destination PE could be wrong
```
Are there any checks on this at runtime? Shouldn't we abort or warn users about this?

minitu replied:
There isn't one currently, but I'll add a check on the receiver side to compare the current PE and the destination PE and abort if they don't match. As of now the user is expected to build charm with `-DCMK_GLOBAL_LOCATION_UPDATE` (which will go into the documentation).

minitu commented:
@nitbhat @stwhite91 Addressed your reviews

minitu commented:
@stwhite91 @nitbhat Just fixed Windows compilation issue caused by pthreads. Please approve again
